habitat/tasks/nav/instance_image_nav_task.py

```diff
@@ -0,0 +1,246 @@
+    """
+
+    image_goals: List[InstanceImageParameters] = attr.ib(
+        default=None, validator=not_none_validator
+    )
+    object_surface_area: Optional[float] = None
+
+
+@registry.register_sensor
+class InstanceImageGoalSensor(RGBSensor):
+    """A sensor for instance-based image goal specification used by the
+    InstanceImageGoal Navigation task. Image goals are rendered according to
+    camera parameters (resolution, HFOV, extrinsics) specified by the dataset.
+
+    Args:
+        sim: a reference to the simulator for rendering instance image goals.
+        config: a config for the InstanceImageGoalSensor sensor.
+        dataset: a Instance Image Goal navigation dataset that contains a
+        dictionary mapping goal IDs to instance image goals.
+    """
+
+    cls_uuid: str = "instance_imagegoal"
+    _current_image_goal: Optional[VisualObservation]
+    _current_episode_id: Optional[str]
+
+    def __init__(
+        self,
+        sim,
+        config: Config,
+        dataset: "InstanceImageNavDatasetV1",
+        *args: Any,
+        **kwargs: Any,
+    ):
+        self._sim = sim
+        self._dataset = dataset
+        super().__init__(config=config)
+        self._current_episode_id = None
+        self._current_image_goal = None
+
+    def _get_uuid(self, *args: Any, **kwargs: Any) -> str:
+        return self.cls_uuid
+
+    def _get_observation_space(self, *args: Any, **kwargs: Any) -> Space:
+        H, W = (
+            next(iter(self._dataset.goals.values()))
+            .image_goals[0]
+            .image_dimensions
+        )
+        return spaces.Box(low=0, high=255, shape=(H, W, 3), dtype=np.uint8)
+
+    def _get_instance_image_goal(
+        self, img_params: InstanceImageParameters
+    ) -> VisualObservation:
+        spec = habitat_sim.CameraSensorSpec()
+        spec.uuid = f"{self.cls_uuid}_sensor"
+        spec.sensor_type = habitat_sim.SensorType.COLOR
+        spec.resolution = img_params.image_dimensions
+        spec.hfov = img_params.hfov
+        spec.sensor_subtype = habitat_sim.SensorSubType.PINHOLE
+        self._sim.add_sensor(spec)
+
+        agent = self._sim.get_agent(0)
+        agent_state = agent.get_state()
+        agent.set_state(
+            AgentState(
+                position=agent_state.position,
+                rotation=agent_state.rotation,
+                sensor_states={
+                    **agent_state.sensor_states,
+                    spec.uuid: SixDOFPose(
+                        position=np.array(img_params.position),
+                        rotation=quaternion_from_coeff(img_params.rotation),
+                    ),
+                },
+            ),
+            infer_sensor_states=False,
+        )
+
+        # render observations from this sensor only
+        self._sim._sensors[spec.uuid].draw_observation()
+        img = self._sim._sensors[spec.uuid].get_observation()[:, :, :3]
+
+        # TODO: delete sensor safely, like sim.remove_sensor(uuid)
```

Comment:
Like we discussed, instead of adding and removing the sensor at each episode, add it once and then keep reusing it.

Reply:
(clarification : this is a suggestion)

Reply:
I looked into it and the python wrapper class around the c++ sensor has a coupling of sim and agent. To add a sensor detached from the agent would require replicating functionality outside of these classes which would be error-prone and rigid. I think its best to use the method as-is and simplify with the eventual sensor binding update.